src/libstore/s3-binary-cache-store.cc

```diff
@@ -184,6 +186,75 @@ S3Helper::FileTransferResult S3Helper::getObject(
     return res;
 }
 
+/* An adapter that can be used to convert a Sink to ostream or iostream.
+
+   Example:
+
+   SinkToBidirectionalDeviceAdapter deviceToSink(sink);
+   boost::iostreams::stream<SinkToBidirectionalDeviceAdapter> streamToSink(deviceToSink);
+
+ */
+struct SinkToBidirectionalDeviceAdapter {
+    typedef char char_type;
+    typedef boost::iostreams::bidirectional_device_tag category;
+    Sink & sink;
+    std::streamsize totalSize;
+
+    SinkToBidirectionalDeviceAdapter(Sink & sink) : sink(sink), totalSize(0) {}
+
+    std::streamsize read(char* data, std::streamsize dataSize)
+    {
+        // Intentionally reports no data read. This struct is only intended for sinks.
+        return 0;
+    }
+
+    std::streamsize write(const char* data, std::streamsize dataSize)
+    {
+        auto strViewData = std::string_view(data, dataSize);
+        this->sink(strViewData);
+        this->totalSize += dataSize;
+        return dataSize;
+    }
+};
+
+S3Helper::FileStreamResult S3Helper::getObjectStreamed(
+    const std::string & bucketName, const std::string & key, Sink & sink)
+{
+    debug("streaming from 's3://%s/%s'...", bucketName, key);
+
+    auto request =
+        Aws::S3::Model::GetObjectRequest()
+        .WithBucket(bucketName)
+        .WithKey(key)
+        .WithResponseContentEncoding("identity");
+
+    SinkToBidirectionalDeviceAdapter streamToSink(sink);
+
+    request.SetResponseStreamFactory([&]() {
+        return Aws::New<boost::iostreams::stream<SinkToBidirectionalDeviceAdapter>>("StreamToSink", streamToSink);
+    });
+
+    S3Helper::FileStreamResult res;
+
+    auto now1 = std::chrono::steady_clock::now();
+
+    try {
+        checkAws(fmt("AWS error streaming '%s'", key), client->GetObject(request));
+        res.streamSuccessful = true;
+        res.size = streamToSink.totalSize;
+    } catch (S3Error & e) {
+        if ((e.err != Aws::S3::S3Errors::NO_SUCH_KEY) &&
+            (e.err != Aws::S3::S3Errors::RESOURCE_NOT_FOUND) &&
```

Comment:
This is another interesting difference from the old `getObject` function. AWS started to respond with `RESOURCE_NOT_FOUND` instead of `NO_SUCH_KEY` when the object does not exist in S3. We do not know why this started happening, but the only difference between our code and the old `getObject` function is the fact that we pin encoding to `identify`, perhaps internally somehow this causes AWS to change its response from `NO_SUCH_KEY` to `RESOURCE_NOT_FOUND`.